GHE g-function calculation enhancements

j-c-cook opened the pull request:
Pull request overview
---------------------

A C++ library, named cpgfunctionEP, with the ability to compute 
g-functions has been developed by Jack C. Cook, a student of Dr. 
Jeffrey D. Spitler at Oklahoma State University.

The cpgfunction library is a low level implementation of Massimo 
Cimmino's g-function methodology. The code was developed with the aim 
of increasing performance (compared to Cimmino's open source Python 
implementation, pygfunction), in both memory and time to be used in 
the generation of g-function databases on high performance computing 
clusters (HPCC). The resulting program makes it more practical to 
compute g-functions for larger ground heat exchangers on a desktop 
computer.

This proposal specifically aims to address issue #6651, to provide
EnergyPlus with an enhanced g-function calculation. The library will be  
located in the third party folder.


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Commit:
Add new feature proposal draft for cpgfunction

A new feature proposal draft for a fast and accurate g-function
calculation is included in this commit. A C++ library, named
cpgfunction, with the ability to compute g-functions has been
developed by Jack C. Cook, a student of Dr. Jeffrey D. Spitler at
Oklahoma State University.

The cpgfunction library is a low level implementation of Massimo
Cimmino's g-function methodology. The code was developed
with the aim of increasing performance (compared to Cimmino's open
source Python implementation, pygfunction), in both memory and time to
be used in the generation of g-function databases on high performance
computing clusters (HPCC). The resulting program makes it more practical
to compute g-functions for larger ground heat exchangers on a desktop
computer.

More information about the g-function calculation can be found in the
NFP.

This proposal specifically aims to address issue #6651, to provide
EnergyPlus with a 3rd party g-function calculation.

Commit:
Refactor NFP mark sample IDF and IDD changes

The NFP mentions proposals for new keys for the
GroundHeatExchanger:System. This would result in new lines in IDF and
IDD sections. Sample IDF and IDD are presented with proposed changes.
It is helpful for the reader to quickly be able to distinguish the
new lines.

The code blocks in markdown do not allow for highlighting (e.g. bold,
underline, etc.). Additionally, html font highlighting did not appear
to render in the .md file on github. The changed lines were placed
below, and were intended to be highlighted in red (rendering md in atom
offline looked fine), but were not. Therefore, the proposed changed
lines now begin with a ">" to show distinction.

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

mitchute commented:
This NFP is in good shape. 

One final comment regarding the NFP: it looks like the gains from the "adaptive integration scheme" are really going to be significant here. You have a plot showing the memory reductions in going from pygfunction to cpgfunction, and then to cpgfunction with the ADS (plot with red arrows). If you have it, it would be nice to show a similar plot for time reductions. This is obviously dependent on the GHE configuration, but an example would be useful. If you don't have it right now, we don't need to hold this up for it.

This is really going to be a nice addition to E+!

FYI: @j-c-cook @jeffreyspitler

jmarrec commented:
Just a tiny note, note that the `diff` language marker can be used for your IDD snippets instead of adding a `>` to mark new lines.

eg: 

```diff
! IDF example using GHE:Vertical:Array input
GroundHeatExchanger:System,
  Vertical GHE 1x4 Std,    !- Name
  GLHE Inlet,              !- Inlet Node Name
  GLHE Outlet,             !- Outlet Node Name
  0.004,                   !- Design Flow Rate {m3/s}
  Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Undisturbed Ground Temperature Model Type
  KATemps,                 !- Undisturbed Ground Temperature Model Name
  2.5,                     !- Ground Thermal Conductivity {W/m-K}
  2.5E+06,                 !- Ground Thermal Heat Capacity {J/m3-K}
  ,                        !- GHE:Vertical:ResponseFactors Object Name
+ UHFcalc,                 !- g-function Calculation Model Name
  GHE-Array;               !- GHE:Vertical:Array Object Name
```

it also works for modifications

```diff
- something old
+ something new
```

nrel-bot-2c commented:
@j-c-cook @lgentile it has been 28 days since this pull request was last updated.

Commit:
Merge remote-tracking branch 'origin/develop' into 6651_cpgfunction_thirdparty

Commit:
Merge branch '6651_cpgfunction_thirdparty' of https://github.com/j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

Commit:
Merge remote-tracking branch 'upstream/develop' into 6651_cpgfunction_thirdparty

Commit:
Append design documentation to NFP

related to #6651

Commit:
Document IDD snippets with the diff language marker

As noted by @jmarrec, there is a way to highlight lines that conain
  - something old
  + something new

https://github.com/NREL/EnergyPlus/pull/8708#issuecomment-823278771

Commit:
Add diff as markdown language for IDD snippets

Commit:
Refactor diff Documentation placements

For the diff highlighting to work on github, the !'s, +'s and -'s must
be at the beginning of a new line.

Commit:
Document the design document in regards to dependencies

relates to #6651

Commit:
Modify design document to reflect outlook

On the EnergyPlus call Wednesday, June 30, 2021, a path was laid out
for the future of cpgfunction. The project has been forked. There now
exists a fork by the name of cpgfunctionEP.

cpgfuncitonEP (a fork of cpgfunction) will contain neither Fortran
nor Boost dependencies.

Commit:
Update design document to reflect current outlook

Commit:
Refactor design document to only include "bottom line up front"

Commit:
Add in links to design document

Commit:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

Commit:
Add cpgfunctionEP-0.1 to the third_party folder

Commit:
Add cpgfunctionEP subdirectory to E+ CMakeLists for third party

Commit:
Comment out cpgfunctionEP tests temporarily

Commit:
Modify GHE.unit.cc GHE:System object for new field

Commit:
Add copy of GSHPSimgple-GLHE.idf file

Commit:
Modify GLHE idf file (copy) to call UBHWT calc

Commit:
Add GHE:CalculationModel name paragraph to IO tex

Commit:
Add gFunctionCalcModelName field to IDF listings

Commit:
Add TODO note in EngrRef for docs location

Commit:
Add TODO note again

Commit:
Add IDD modifications (memo and A7)

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

mitchute committed:
Merge branch '6651_cpgfunction_thirdparty' of github.com:j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

mitchute committed:
Revert "Add IDD modifications (memo and A7)"

This reverts commit e8f92b64cdb215c8bf469eb33128aab2ac0355d9.

mitchute committed:
modify ghe inputs for new calc options, add cpgfunction build to alt-build, update test files and unit tests with new input args

Commit:
Remove cpgfunctionEP-v0.1

Commit:
Add cpgfunction-v0.2

Commit:
Merge branch '6651_cpgfunction_thirdparty' of https://github.com/j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

mitchute committed:
integrate new cpgfunction version into E+ which is now generating gfunctions from the new UBWT code

Commit:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

Commit:
Increment cpgfunctionEP to version 0.3

Commit:
Modify third party cmakelists to reference cpg0.3

Commit:
Remove unnecessary find Eigen library linker cmake

Commit:
Make eigen available for cpgfunctionEP in cmakelists

Commit:
Add Spitler's first version of engr ref

mitchute committed:
integrate latest cpgfunction update, update ghe code to properly switch algorithms and compute lts and sts g-functions, cleanup warnings

Commit:
Merge branch '6651_cpgfunction_thirdparty' of https://github.com/j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

Commit:
Add adaptively discretized number of segments

The solution to the ubhwt g-function is dependent on the number of
segments used. The solution defaults to 12, however the adaptive
discretization provides a number of segments that will give a
g-function close to a converged UIFT g-function.

Commit:
Reformat ghe.cc to pass clang-format test

Commit:
Reformat ghe.cc (again) topass clang format check

Commit:
Fix call to adaptive discretization (clang-tidy)

Commit:
Document heat-exchangers.tex about segments

The UBHWTCalc makes use of an adaptive discretization scheme to
approximate a UIFT g-function.

Commit:
Document heat-exchangers.tex with new references

Commit:
Documentation of ghe

Commit:
Modify g-function all UBHWTCalc->UBHWTcalc

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

mitchute committed:
add unit test, correct incorrect key in example file, add new methods to caching checks, minor cleanups

mitchute committed:
clang-format

Commit:
Add new cpgfunctionEP files

Commit:
Bump cpgfunctionEP to next version

Commit:
Modify main cmakelists to reference new version

Commit:
Modify third party cmakelists for new version

Commit:
Modify time declaration in E+ for call to cpgEP

The g-function methodology of Cimmino can become flawed if the time
vector is poorly chosen.

Commit:
Merge branch '6651_cpgfunction_thirdparty' of https://github.com/j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

mitchute committed:
Merge branch '6651_cpgfunction_thirdparty' of github.com:j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

Commit:
Revert "Modify time declaration in E+ for call to cpgEP"

This reverts commit c2a539c14c81ff345beaf5ab585c75f798f6530e.

Commit:
Bump cpgfunctionEP to 0.3.2

Commit:
Rename cpgfunctionEP-0.3.1->cpgfunctionEP-0.3.2

Commit:
Bump cpgfunction references in cmakelists

mitchute committed:
Merge branch '6651_cpgfunction_thirdparty' of github.com:j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

mitchute committed:
update unit test

mitchute committed:
update docs

mitchute committed:
update docs

mitchute committed:
cleanup eol null chars to eliminate build warnings

mitchute commented on .github/workflows/linux_build.yml:
```diff
@@ -69,6 +69,7 @@ jobs:
                 -DUSE_PSYCH_STATS=ON \
                 -DUSE_PSYCH_ERRORS=OFF \
                 -DENABLE_PCH=OFF \
+                -DBUILD_CPGFUNCTION=ON \
```
Whoops, that needs to go away. I'll fix that on the next commit.

mitchute commented on CMakeLists.txt:
```diff
@@ -265,6 +265,9 @@ target_include_directories(project_options SYSTEM INTERFACE "${kiva_SOURCE_DIR}/
 set_target_properties(generate_embeddedable_epJSON_schema PROPERTIES FOLDER "Internal")
 
+target_include_directories(project_options INTERFACE ${PROJECT_SOURCE_DIR}/third_party/cpgfunctionEP-0.3.2)
+target_include_directories(project_options INTERFACE ${PROJECT_SOURCE_DIR}/third_party/cpgfunctionEP-0.3.2/include)
```
New CMake rules for the new library.

mitchute commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-002/heat-exchangers.tex:
```diff
@@ -929,7 +929,11 @@ \subsubsection{Long Time-Step Response Factors}\label{long-time-step-response-fa
 The \emph{g}-functions developed by Eskilson are given only down to values of $\ln t/t_s = -4.5$, which for typical system and soil types can be on the order of 100 days. The values can be extended by the applying a line source model down to a time of $t = 5r_b^2/\alpha$. This time varies from 3-6 hours for a typical borehole field. This is because the analytical line source model, based on which ...
 
-In order to generate the long time-step response factors on the fly, EnergyPlus uses the model developed by Marcotte \& Pasquier(2009) which uses a discretized line source model. The \emph{g}-functions are generated using the following equation. The boreholes are discretized into segments. The temperature response of each segment on all other segments is then used to determine response factor for ...
+In order to generate the long time-step response functions on the fly, EnergyPlus has two different models that make two different approximations for the calculation of the response functions. The first model uses uniform heat flux boundary conditions (UHFcalc model) – that is, it assumes that all boreholes have the same uniform heat flux. The heat flux may vary over time, but it is always unifor ...
```
These Eng Ref changes are a nice contribution to this section.

mitchute commented on doc/input-output-reference/src/overview/group-condenser-equipment.tex:
```diff
@@ -3026,6 +3026,10 @@ \subsubsection{Inputs}\label{inputs-10-002}
 
+\paragraph{Field: g-Function Calculation Method}
+
+If the GHE:ResponseFactors object is not provided, this input will be relevant to the calculation method used to generate the g-function values. There are two options for this input: ``UHFcalc" or ``UBHWTcalc". The default is UHFcalc, where a g-function is computed with the uniform heat flux boundary condition. The UBHWTcalc option computes the g-function with a uniform borehole wall temperature  ...
```
New I/O Ref changes for the new field.

mitchute commented on idd/Energy+.idd.in:
```diff
@@ -74798,308 +74799,313 @@ GroundHeatExchanger:System,
+        \key UHFcalc
+        \key UBHWTcalc
+        \default UHFcalc
+  A8,   \field GHE:Vertical:Array Object Name
```
New IDD description of this new field.

mitchute commented on src/EnergyPlus/GroundHeatExchangers.cc:
```diff
@@ -318,49 +324,69 @@ GLHEVert::GLHEVert(EnergyPlusData &state, std::string const &objName, nlohmann::
+    if (!this->gFunctionsExist) {
+
+        // g-function calculation method
+        if (j.find("g_function_calculation_method") != j.end()) {
```
Input changes to allow for the new field.

mitchute commented on src/EnergyPlus/GroundHeatExchangers.cc:
```diff
@@ -946,13 +972,53 @@ Real64 GLHEVert::doubleIntegral(std::shared_ptr<GLHEVertSingle> const &bh_i, std
+    default:
+        assert(false);
+    }
+}
```
There was also some general refactoring to allow switching on these two methods.

mitchute commented on testfiles/CMakeLists.txt:
```diff
@@ -296,6 +296,7 @@ add_simulation_test(IDF_FILE FurnaceWithDXSystemRHcontrol_cyclingfan.idf EPW_FIL
 add_simulation_test(IDF_FILE FurnaceWithDXSystem_CoolingHXAssisted.idf EPW_FILE USA_FL_Miami.Intl.AP.722020_TMY3.epw)
 add_simulation_test(IDF_FILE GSHP-GLHE.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE GSHP-GLHE-CalcGFunctions.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+add_simulation_test(IDF_FILE GSHP-GLHE-CalcGFunctions-cpg.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
```
New test case with the new library and methods.

mitchute commented on third_party/cpgfunctionEP-0.3.2/.gitignore:
```diff
@@ -0,0 +1,7 @@
+cmake-build-debug/
```
When viewing the files changed on GitHub, everything from here down is the new library. I'm not going to spend too much time in there reviewing, other than to note that most of these files are due to the library bringing in a stripped-down version of Eigen.

mitchute reviewed (commented):
Code walkthrough.

mitchute committed:
remove old build flag

j-c-cook commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-002/heat-exchangers.tex:
```diff
@@ -929,7 +929,11 @@ \subsubsection{Long Time-Step Response Factors}\label{long-time-step-response-fa
-In order to generate the long time-step response factors on the fly, EnergyPlus uses the model developed by Marcotte \& Pasquier(2009) which uses a discretized line source model. The \emph{g}-functions are generated using the following equation. The boreholes are discretized into segments. The temperature response of each segment on all other segments is then used to determine response factor for ...
+In order to generate the long time-step response functions on the fly, EnergyPlus has two different models that make two different approximations for the calculation of the response functions. The first model uses uniform heat flux boundary conditions (UHFcalc model) – that is, it assumes that all boreholes have the same uniform heat flux. The heat flux may vary over time, but it is always unifor ...
+
+Calculation of \emph{g}-functions with uniform heat flux boundary conditions is simpler, but accuracy can decrease as the borehole-to-borehole interference increases. Consider a case with a large rectangular borefield, say 9x16 boreholes – if there is a significant annual imbalance in the heat rejection/heat extraction, the interior of the field will become thermally saturated, and the heat flux  ...
```
Discusses UHF being less accurate, provides reference to paper that shows an imbalance in heat rejection/heat extraction when UBHWT is used.

j-c-cook commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-002/heat-exchangers.tex:
```diff
@@ -953,6 +957,11 @@ \subsubsection{Long Time-Step Response Factors}\label{long-time-step-response-fa
 
 \(d(u_i, u_j^{'})\) is the distance between current point and point on other imaginary borehole.
 
+For the UBHWTcalc model, EnergyPlus builds on the model developed by Cimmino (2018a, 2018b, 2019), with improvements (Cook and Spitler 2021) to reduce memory consumption and increase speed. The general approach is similar to that of the UHFcalc model, but the heat flux is adjusted for each segment to enforce the uniform borehole wall temperature model. Additionally, the methodology of Cimmino (UB ...
```
Describes UBHWTcalc as superior to UHFcalc and discusses UBHWTcalc's approximation of UIFT.

j-c-cook commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-002/heat-exchangers.tex:
```diff
@@ -1129,15 +1138,57 @@ \subsubsection{Summary of Variable Short Time Step Response Factor Model}\label{
 
 \subsubsection{References}\label{references-2-006}
 
-Eskilson, P. 1987. Thermal Analysis of Heat Extraction Boreholes. Ph.D.~Thesis, Department of Mathematical Physics, University of Lund, Lund, Sweden.
```
Many new references. References reformatted to have a hanging indent beginning at the second line. Looks much cleaner.

j-c-cook commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-002/heat-exchangers.tex:
```diff
@@ -1129,15 +1138,57 @@ \subsubsection{Summary of Variable Short Time Step Response Factor Model}\label{
+
+\hangindent=2em
+\hangafter=1
+\noindent Cimmino, M. (2015). "The effects of borehole thermal resistances and fluid flow rate on the g-functions of geo-thermal bore fields." International Journal of Heat and Mass Transfer 91: 1119-1127.
```
Not sure what these red highlights are in Cimmino's citations. I'll see if I can get those off on my next push.

mjwitte replied:
It doesn't like the double-quote symbol. Quotes should open with two back-ticks (I forget the proper name) and close with two apostrophes. ``set up''

j-c-cook replied:
Looks like that resolved it, thanks @mjwitte !! b68c587f7080c9e7bfd9afe226c5dbd22d4dee5d

Commit:
Remove Cimmino's references to create diff

Commit:
Add Cimmino's references as plain text

There was some red highlighting of the titles on githubs diff markup.
This is being done to make it all be plain text.

j-c-cook commented on src/EnergyPlus/GroundHeatExchangers.cc:
```diff
@@ -946,13 +972,53 @@ Real64 GLHEVert::doubleIntegral(std::shared_ptr<GLHEVertSingle> const &bh_i, std
+        assert(false);
+    }
+}
+
```
The UBHWTcalc makes use of the same borehole locations and time steps as UHFcalc would, they just need stored in a vector first.

j-c-cook commented on src/EnergyPlus/GroundHeatExchangers.cc:
```diff
@@ -946,13 +972,53 @@ Real64 GLHEVert::doubleIntegral(std::shared_ptr<GLHEVertSingle> const &bh_i, std
+    for (auto &v : this->myRespFactors->time) {
+        time.push_back(v);
+    }
+
```
The adaptive discretization struct is called to choose a number of segments for UBHWTcalc that will approximate a UIFT g-function.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/CHANGELOG.md:
```diff
@@ -0,0 +1,211 @@
+# History of changes - cpgfunctionEP
```
A markdown file that contains a history of changes (CHANGELOG) for cpgfunctionEP and the project prior to a fork, cpgfunction. The changes are organized by release and category.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+cmake_minimum_required(VERSION 3.6)
```
The CMake lists reference local libraries only. There is nothing additional a user needs to have installed to make use of UBHWTcalc. If the project is built by E+ the internal tests will not be built. The internal test files and library (nlohmann json) are not included in this copy of the library.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/blas/blas.h:
```diff
@@ -0,0 +1,26 @@
+//
```
Two BLAS functions were ported to C++, `axpy` and `spmv`. 

axpy - takes the form `y = ax + y`
spmv - takes the form `y = Ax + y` where A is a packed symmetric matrix

j-c-cook replied:
The following are the 18 project files associated with cpgfunctionEP.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/boreholes.h:
```diff
@@ -0,0 +1,96 @@
+//
```
Header file for boreholes. All things related to borehole properties and positions are in this module.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/CONTRIBUTING.md:
```diff
@@ -0,0 +1,125 @@
+# Contributing to cpgfunction
```
This is a contributing file for cpgfunctionEP that are good for new contributors and helpful for myself to look back at. This doesn't need to be included for E+, will be removed on the next push.

Commit:
Remove CONTRIBUTING.md file

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/README.md:
```diff
@@ -0,0 +1,88 @@
+# cpgfunctionEP
```
The README is useful and relevant information.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/LICENSE:
```diff
@@ -0,0 +1,29 @@
+BSD 3-Clause License
```
BSD 3 License.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/coordinates.h:
```diff
@@ -0,0 +1,35 @@
+//
```
This file contains functions that create coordinates for typical configurations: rectangle, open rectangle, U, and L.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/gfunction.h:
```diff
@@ -0,0 +1,63 @@
+// -*- lsst-c++ -*-
```
Functions related to the g-function calculation methodology described in Cimmino and Bernier (2014).

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/heat_transfer.h:
```diff
@@ -0,0 +1,28 @@
+//
```
Functions to compute the finite line source solutions. Ie. segment-to-segment response matrix.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/interpolation.h:
```diff
@@ -0,0 +1,25 @@
+//
```
User defined interpolation functions made use of throughout the library.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/segments.h:
```diff
@@ -0,0 +1,85 @@
+//
```
Given that so much work was put into the segment responses (packing the matrix and the adaptive discretization scheme), there is a module related to handling segments.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/statistics.h:
```diff
@@ -0,0 +1,20 @@
+//
```
Statistics module that is used in running tests. Contains RMSE percentage function.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/cpgfunction/utilities.h:
```diff
@@ -0,0 +1,41 @@
+//
```
These are functions associated with creating a time vector for input to the g-function calculation. None of these are actually used in E+ since Matt has his own constant time expansion time declaration function.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/src/blas.cpp:
```diff
@@ -0,0 +1,43 @@
+//
```
I've described what the files are at the header (.h) files, and will not do the same for the source files (.cpp).

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/third_party/eigen-3.3.9/.gitignore:
```diff
@@ -0,0 +1,37 @@
+qrc_*cxx
```
Will be removed from Eigen lib.

Commit:
Remove uneccessary hidden . files from eigen

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/third_party/eigen-3.3.9/Eigen/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+include(RegexUtils)
```
Everything below here is eigen related. The library was stripped. The total number of files related to eigen went from 302 to 146. 

The following are teh 146 files associated with the `third_party/eigen` library.

jmarrec replied:
Seems like we'll now be carrying three copies of eigen in our third_party directory. Could we reuse eigen instead? or am I missing something?

j-c-cook replied:
@Myoldmopar @mitchute It looks like egien-3.3.7 is locaed in `third_party/eigen`. Would it be okay for me to link to that copy of eigen?

Myoldmopar replied:
It would be great if we could link them all to the same Eigen, but the problem is that as we use more third_party tools, there is a higher chance that one of them will require a specific version of Eigen.  So linking them together then becomes a burden if you have to break them apart again later.  Since our third_party tools don't change frequently, and if you can get them to work with our already included Eigen, that would be very very nice.  But I won't hold this up for it.

j-c-cook replied:
Yes sir, I'll see what I can do. Thank you.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/third_party/qdt-master/.gitignore:
```diff
@@ -0,0 +1,28 @@
+# Compiled Object files
```
This can be removed on the next push.

Commit:
Remove .gitignore file from qdt

Commit:
Merge branch '6651_cpgfunction_thirdparty' of https://github.com/j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/third_party/qdt-master/README.md:
```diff
@@ -0,0 +1,67 @@
+# qdt
```
The `qdt` library, written by a masters student in 2006. Adolfo Munoz wrote the guassian kronrod integration for a paper he published, and graciously made the functions open source. His library enabled me to get rid of the boost function call to `boost::math::quadrature::kronrod`

j-c-cook commented on tst/EnergyPlus/unit/GroundHeatExchangers.unit.cc:
```diff
@@ -1333,6 +1336,360 @@ TEST_F(EnergyPlusFixture, GroundHeatExchangerTest_System_calcGFunction_Check)
     EXPECT_NEAR(thisGLHE.interpGFunc(-3.963), 5.82, tolerance);
 }
 
+TEST_F(EnergyPlusFixture, GroundHeatExchangerTest_System_calcGFunction_UBHWT)
```
EnergyPlus unit test with UBHWTcalc.

j-c-cook commented on tst/EnergyPlus/unit/GroundHeatExchangers.unit.cc:
```diff
@@ -862,8 +862,9 @@ TEST_F(EnergyPlusFixture, GroundHeatExchangerTest_System_Given_Array_IDF_Check)
-                          "    ,                    !- Response Factors Object Name",
-                          "    GHE-Array;            !- !- GHE Array Object Name"});
+                          "    ,                   !- Response Factors Object Name",
+                          "    UHFCalc,            !- g-Function Calculation Method",
```
Adds new IDD field for UHFcalc or UBHWTcalc.

j-c-cook commented on tst/EnergyPlus/unit/GroundHeatExchangers.unit.cc:
```diff
@@ -1333,6 +1336,360 @@ TEST_F(EnergyPlusFixture, GroundHeatExchangerTest_System_calcGFunction_Check)
+
+    Real64 const tolerance = 0.1;
+
+    // Test g-function values from GLHEPro
```
If this is truly testing the g-function values from GLHEPRO, we may have an issue if the field is large enough. For smaller fields this won't be a problem, but the UIFT approximation becomes quite a bit different (as much as +15%) when fields become larger. Have you made sure this wont break @mitchute ?

j-c-cook commented on src/EnergyPlus/GroundHeatExchangers.cc:
```diff
@@ -946,13 +972,53 @@ Real64 GLHEVert::doubleIntegral(std::shared_ptr<GLHEVertSingle> const &bh_i, std
+
+//******************************************************************************
+
+void GLHEVert::calcUniformBHWallTempGFunctions()
+{
+    // construct boreholes vector
+    std::vector<gt::boreholes::Borehole> boreholes;
+    for (auto &bh : this->myRespFactors->myBorholes) {
+        boreholes.emplace_back(bh->props->bhLength, bh->props->bhTopDepth, bh->props->bhDiameter / 2.0, bh->xLoc, bh->yLoc);
+    }
+
+    // convert time to a std::vector from an Array1D
+    std::vector<double> time;
+    for (auto &v : this->myRespFactors->time) {
+        time.push_back(v);
+    }
+
+    // Obtain number of segments by adaptive discretization
+    gt::segments::adaptive adptDisc;
+    int nSegments = adptDisc.discretize(this->bhLength, this->totalTubeLength);
+
+    this->myRespFactors->GFNC = gt::gfunction::uniform_borehole_wall_temperature(boreholes, time, this->soil.diffusivity, nSegments);
+}
```
The UBHWTcalc makes use of vector rather than Array1D.

The adaptive discretization struct is called to choose a number of segments for UBHWTcalc that will approximate a UIFT g-function.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/LICENSE:
```diff
@@ -0,0 +1,29 @@
+BSD 3-Clause License
+
+Copyright (c) 2020, Jack C. Cook
+All rights reserved.
```
BSD-3 license.

j-c-cook commented on third_party/cpgfunctionEP-0.3.2/include/blas/blas.h:
```diff
@@ -0,0 +1,26 @@
+
+namespace jcc  { namespace blas {
+
+        void axpy(int &n, double &a, vector<double> &x, vector<double> &y,
+                  int &start, int &n_threads);
+
+        void spmv(int &n, double &alpha, vector<double> &A, vector<double> &x,
+                  double &beta, vector<double> &y, int &start, int &n_threads);
```
Two BLAS functions were ported to C++, `axpy` and `spmv`. 

axpy - takes the form `y = ax + y`
spmv - takes the form `y = Ax + y` where A is a packed symmetric matrix

Commit:
Fix double qoute issue in Cimmino's references

thanks @mjwitte for your input on how to resolve the red highlights in
Cimmino's citations.

mitchute commented:
While this is under review, I want to point out that the new library is set up to multithread when computing the g-functions. It's currently set to consume all threads during this calculation. See here:

https://github.com/j-c-cook/EnergyPlus/blob/b68c587f7080c9e7bfd9afe226c5dbd22d4dee5d/third_party/cpgfunctionEP-0.3.2/include/cpgfunction/gfunction.h#L42

https://github.com/j-c-cook/EnergyPlus/blob/b68c587f7080c9e7bfd9afe226c5dbd22d4dee5d/third_party/cpgfunctionEP-0.3.2/src/gfunction.cpp#L97-L100

Should we expose this to the user as a parameter in the GHE object so they can control it, or does it need to live at a higher level in E+ inputs? Or do we just not worry about it at this point?

jmarrec commented on third_party/cpgfunctionEP-0.3.2/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+
+set(CMAKE_CXX_STANDARD 17)
+if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --std=c++17")
```
`set(CMAKE_CXX_STANDARD 17)` will take care of adding `-std=c++17`, and here you have an extra `-` anyways too.

j-c-cook replied:
Believed to be addressed in a7f66c005851db3116f729a4ad814d4d86fd0d41

jmarrec commented on third_party/cpgfunctionEP-0.3.2/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --std=c++17")
+else()
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
https://cmake.org/cmake/help/v3.4/variable/CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS.html

`CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` is the default value for `WINDOWS_EXPORT_ALL_SYMBOLS` which itself is "only implemented for MS-compatible tools on Window".

Perhaps the block should just be this?

```
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
if (MSVC)
  add_compile_definitions(_USE_MATH_DEFINES)
endif()
```

j-c-cook replied:
Believed to be addressed in a7f66c005851db3116f729a4ad814d4d86fd0d41

jmarrec commented on third_party/cpgfunctionEP-0.3.2/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+include_directories(third_party/qdt-master)
+
+# Create a DLL with an API
+add_library(cpgfunctionEP SHARED
```
do we have a need for this to be explicitly SHARED? If the goal is to always link it to energypluslib, then STATIC would make more sense. But I would omit the keyword, rely on the value of `BUILD_SHARED_LIBS` to determine if STATIC or SHARED is picked.

j-c-cook replied:
Believed to be addresssed in a7f66c005851db3116f729a4ad814d4d86fd0d41

jmarrec commented on third_party/cpgfunctionEP-0.3.2/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+
+# It depends on threads
+find_package(Threads REQUIRED)
+target_link_libraries(cpgfunctionEP ${CMAKE_THREAD_LIBS_INIT})
```
Modernize the target_link_libraries by explcitly passing the type (PUBLIC, PRIVATE, INTERFACE). Here it's just passing `-lthreads` basically, so it's an INTERFACE.

```suggestion
target_link_libraries(cpgfunctionEP INTERFACE Threads::Threads)
```

jmarrec commented on third_party/cpgfunctionEP-0.3.2/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+find_package(Threads REQUIRED)
+target_link_libraries(cpgfunctionEP ${CMAKE_THREAD_LIBS_INIT})
+
+OPTION (USE_OpenMP "Use OpenMP" ON)
+IF(USE_OpenMP)
+    FIND_PACKAGE(OpenMP)
+    IF(OPENMP_FOUND)
+        SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+    ENDIF()
+ENDIF()
```
I am not a big fan of messing with CMAKE_C_FLAGS/CXX_FLAGS. The canonical way to do this is:

```
find_package(OpenMP)
if(OpenMP_CXX_FOUND)
    target_link_libraries(MyTarget PUBLIC OpenMP::OpenMP_CXX)
endif()
```

**I think `PRIVATE` linking is appropriate here, maybe it's `PUBLIC` though.**

```suggestion
option (USE_OpenMP "Use OpenMP" ON)
if(USE_OpenMP)
    find_package(OpenMP)
    if(OpenMP_CXX_FOUND)
      target_link_libraries(cpgfunctionEP PRIVATE OpenMP::OpenMP_CXX)
    else()
      message(AUTHOR_WARNING "OpenMP_CXX not found")
    endif()
endif()
```

j-c-cook replied:
Looks like private works.

jmarrec commented on third_party/cpgfunctionEP-0.3.2/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+    add_executable(packed test/packed.cpp)
+    add_executable(gFunction_adaptive test/gFunction_adaptive.cpp)
+
+    target_link_libraries(gFunction_minimal cpgfunctionEP)
+    target_link_libraries(interpolation cpgfunctionEP)
+    # target_link_libraries(test3 cpgfunctionEP)
+    target_link_libraries(borefield_definition cpgfunctionEP)
+    target_link_libraries(time_definition cpgfunctionEP)
+    target_link_libraries(gFunction_tests cpgfunctionEP)
+    target_link_libraries(packed cpgfunctionEP)
+    target_link_libraries(gFunction_adaptive cpgfunctionEP)
```
```suggestion
    target_link_libraries(gFunction_minimal PRIVATE cpgfunctionEP)
    target_link_libraries(interpolation PRIVATE cpgfunctionEP)
    # target_link_libraries(test3 PRIVATE cpgfunctionEP)
    target_link_libraries(borefield_definition PRIVATE cpgfunctionEP)
    target_link_libraries(time_definition PRIVATE cpgfunctionEP)
    target_link_libraries(gFunction_tests PRIVATE cpgfunctionEP)
    target_link_libraries(packed PRIVATE cpgfunctionEP)
    target_link_libraries(gFunction_adaptive PRIVATE  cpgfunctionEP)
```

jmarrec reviewed (commented):
I only did a review of the CMake part.

Commit:
Bump cpgfunctionEP folder name to 0.4

Commit:
Bump cpgfunctionEP files to 0.4

The changes from version 0.3.2 to version 0.4 of cpgfunctionEP are
based on @jmarrec's expertise about cmake in regards to policy and
best practices.

Commit:
Remove local copy of Eigen library

Based on @jmarrec's recommendation, the Eigen library located in
EnergyPlus's third party folder will be referenced rather than
including a shrunk copy of Eigen with this library. Even with the
reduced library that was being included, there were many files. If
this manages to build on all OS's successfully, it will make the merge
diffs look much nicer and more tractable.

Commit:
Make EP cmake link to bumped cpgfunctionEP folder

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

mitchute committed:
Merge branch '6651_cpgfunction_thirdparty' of github.com:j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

mitchute commented:
I'm doing some local testing on this PR this morning, but I think we're close here. 

@Myoldmopar @jmarrec @mjwitte please see [this comment](https://github.com/NREL/EnergyPlus/pull/8708#issuecomment-889513763) re: the ability of this new library to multi-thread. We should come to some consensus on that issue before we merge this PR.

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

mitchute commented:
I ran tests locally and they are consistent with what's coming out of CI for the most recent results. I pulled develop in one more time to make sure. Table diffs are due to adding a new default field, which corresponds to the new input field in the GHE:System object. The results from the new library are consistent with the existing g-function generation method that's currently embedded in E+.

It was suggested to me outside of GitHub that we add user control of the multi-threading as a new CLI option rather than a new field in the GHE object. That would likely need to be pulled back out later when E+ or other models begin to take advantage of multi-threading anyway. If that's the direction we decide to go, we'd need to settle on the default value. That'd probably just be "1" or to single-thread, I'm assuming.

Myoldmopar committed:
Merge remote-tracking branch 'upstream/develop' into 6651_cpgfunction_thirdparty

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

Commit:
Bump Eigen library to 3.3.9

Commit:
Bump cpgfunctionEP to 0.4.1

Commit:
Make cmake files link to cpgEP-0.4.1

mitchute committed:
Merge branch '6651_cpgfunction_thirdparty' of github.com:j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

j-c-cook commented:
There is a segmentation fault that is occurring inside of the `Eigen` library upon the API call to `cpgfunctionEP` via the `UBHWTCalc` input on the Linux CI machines and inside of CLion on my local. This segmentation fault has been around and unnoticed for quite some time. The segmentation fault does not exist if I build energyplus from source, and then pass a weather file and the `GSHP-GLHE-CalcGFunctions-cpg.idf` test file. The segmentation fault does not exist when I build `cpgfunctionEP` in my own repository as a standalone. 

```
mkdir build && cd build && cmake .. && make energyplus -j 16
cp ../testfiles/GSHP-GLHE-CalcGFunctions-cpg.idf Products/ && cp ../weather/USA_OK_Oklahoma.City-Will.Rogers.World.AP.723530_TMY3.epw Products/
cd Products/ 
./energyplus -w USA_OK_Oklahoma.City-Will.Rogers.World.AP.723530_TMY3.epw GSHP-GLHE-CalcGFunctions-cpg.idf
```
The resulting long time-step g-function located in the `eplusout.glhe` file:
```
        2.5227814172726757,
        2.7705529489647103,
        3.0181546610457,
        3.2655977297246355,
        3.516846551928096,
        3.7912271892255647,
        4.127136395615098,
        4.56335062650864,
        5.116250608380114,
        5.774475535544771,
        6.509681059546577
```

I have gone to the `j-c-cook/cpgfunctionEP` repository and created a test file that performs the same function call to the `uniform_borehole_wall_temperature()` function. That example is below (Note: I have modified the number of segmnents in the EnergyPlus call to be 1 to reduce the total number of sources so the matrices are more palpable):

```
//
// Created by jackcook on 8/16/21.
//

#include <cpgfunction/coordinates.h>
#include <cpgfunction/boreholes.h>
#include <cpgfunction/utilities.h>
#include <cpgfunction/gfunction.h>


int main() {
    // -- Definitions --
    // Coordinate geometry
    int Nx = 2;
    int Ny = 2;
    double Bx = 5.;
    double By = 5;

    // -- Configurations --
    // Get x,y coordinates for a rectangle
    std::string shape = "Rectangle";
    std::vector<std::tuple<double, double>> coordinates =
            gt::coordinates::configuration(shape, Nx, Ny, Bx, By);

    // -- Borehole geometry --
    double H = 100;  // height of the borehole (in meters)
    double D = 1;  // burial depth (in meters)
    double r_b = 0.057;  // borehole radius (in meters)

    // -- boreField --
    std::vector<gt::boreholes::Borehole> boreField =
            gt::boreholes::boreField(coordinates, r_b, H, D);

    // Thermal properties
    double alpha = 1.0e-6;      // Ground thermal diffusivity (m2/s)

    // Number of segments per borehole
    int nSegments = 1;

    double duration = 1;
    std::string units = "year";
    double const expansion = 0.5;

    vector<double> time = gt::utilities::time_vector_constant_expansion(
            H, alpha, duration, units, expansion);

    std::vector<double> gFunction =
            gt::gfunction::uniform_borehole_wall_temperature(boreField,
                                                             time,
                                                             alpha,
                                                             nSegments,
                                                             true);
    for (size_t i=0; i<gFunction.size(); i++) {
        std::cout << gFunction[i] << std::endl;
    }

}
```
Outputs (Note the exact match out to 16 decimal places to the ./energyplus UBHWT g-function above):
```
2.522781417272676
2.77055294896471
3.0181546610457
3.265597729724635
3.516846551928096
3.791227189225565
4.127136395615098
4.56335062650864
5.116250608380114
5.774475535544771
6.509681059546577
```

The segmentation fault in EnergyPlus occurs upon the first partial pivot LU decomposition API call. The A and B matrix are shown below:

----------------------
A (EnergyPlus -> cpgfunctionEP)
```
    2.52278 1.71313e-14 1.71313e-14 8.60119e-27          -1
1.71313e-14     2.52278 8.60119e-27 1.71313e-14          -1
1.71313e-14 8.60119e-27     2.52278 1.71313e-14          -1
8.60119e-27 1.71313e-14 1.71313e-14     2.52278          -1
        100         100         100         100           0
```
B (EnergyPlus -> cpgfunctionEP)
```
 -0
 -0
 -0
 -0
400
```
---------------------
A (cpgfunctionEP)
```
    2.52278 1.71313e-14 1.71313e-14 8.60119e-27          -1
1.71313e-14     2.52278 8.60119e-27 1.71313e-14          -1
1.71313e-14 8.60119e-27     2.52278 1.71313e-14          -1
8.60119e-27 1.71313e-14 1.71313e-14     2.52278          -1
        100         100         100         100           0
```
B (cpgfunctionEP)
```
 -0
 -0
 -0
 -0
400
```

The A and B matrices appear identical. The g-function results of the `energyplus` executable are identical to `cpgfunctionEP`. The reasoning for the segmentation fault in Linux on the CI machines when running tests, when running tests (ctest) on my local and when running the input file on CLion are unknown. 

Given that this is a timely manner due to the deadline for API changes for the next release to be finalized by this Friday, I am considering replacing the Eigen LU partial pivot decomposition call with a hand written one or finding a standalone.

mitchute committed:
add -j CLI option and wire it up to the cpgfunction code. add notes and other descriptions describing limitations of new -j CLI option

mitchute commented:
FYI @Myoldmopar @mjwitte the new CLI option discussed in [this comment](https://github.com/NREL/EnergyPlus/pull/8708#issuecomment-894385703) was added here: https://github.com/NREL/EnergyPlus/pull/8708/commits/378a2547af8ba9301e95df85c5f4edd095cb5fe8

mitchute commented:
@j-c-cook The `B` matrices from the examples you show in the latest comment _are_ different. What am I missing here?

j-c-cook commented:
> @j-c-cook The `B` matrices from the examples you show in the latest comment _are_ different. What am I missing here?

That was a mistake in the copy and paste. I edited the post.

Commit:
Revert "Bump Eigen library to 3.3.9"

This reverts commit 33a28cd8b81991fadc680f34157ae704aea3b8ec.

Commit:
Bump cpgfunctionEP to 0.5

Commit:
Make EP's cmake point to cpgfunction-v0.5

Commit:
Merge branch '6651_cpgfunction_thirdparty' of https://github.com/j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

mitchute commented:
I've run this branch on two different Linux boxes, in release and debug modes. It looks like the segfaults are now gone. @Myoldmopar over to you for a final review and merge.

Myoldmopar committed:
Merge branch '6651_cpgfunction_thirdparty' of github.com:j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

Myoldmopar committed:
Merge remote-tracking branch 'upstream/develop' into 6651_cpgfunction_thirdparty

Myoldmopar commented:
Building now, thanks @mitchute !

Myoldmopar commented:
Concur, everything is passing now.  I'll be happy to resolve the conflict and do final review here.

Myoldmopar committed:
Add clarification in cli usage

mitchute committed:
Merge branch 'develop' into 6651_cpgfunction_thirdparty

Myoldmopar commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -210,6 +215,18 @@ namespace CommandLineInterface {
+        if (state.dataGlobal->numThread == 0) {
+            DisplayString(state, "Invalid value for -j arg. Defaulting to 1.");
+            state.dataGlobal->numThread = 1;
+        } else if (state.dataGlobal->numThread > (int)std::thread::hardware_concurrency()) {
```
Interesting.  I'm not sure we really need to take the burden of restricting this.  And on some platforms where this cannot be determined, it is required to return 0, thus making EnergyPlus fail.  I think I'll take this out.

Myoldmopar commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -149,6 +152,8 @@ namespace CommandLineInterface {
 
         opt.add("", false, 0, 0, "Display version information", "-v", "--version");
 
+        opt.add("1", false, 1, 0, "Multi-thread with N threads; 1 thread with no arg.", "-j", "--jobs");
```
I'm adding a little phrase in here to describe the GHE usage so it shows up in the usage() display.

Myoldmopar reviewed (commented):
This looks really close.  I noted a few minor tweaks that I'm going to make in a follow-up commit.  This also needs a minor transition rule which @mitchute is doing now.  Otherwise this should be ready to go in.

mitchute committed:
add ghe transition rules

Myoldmopar committed:
Merge branch '6651_cpgfunction_thirdparty' of github.com:j-c-cook/EnergyPlus into 6651_cpgfunction_thirdparty

Myoldmopar commented:
Thanks for the transition update @mitchute.  I took a GSHP file that did not have the full list of fields and it left the object alone.  I took GSHP-GLHE-CalcGFunctions.idf and transitioned it to find the new default field name put in there.  It's working as expected.  I have a few small tweaks and will do final testing, but I anticipate this is good to go in next/shortly.

Myoldmopar committed:
Merge remote-tracking branch 'upstream/develop' into 6651_cpgfunction_thirdparty

Myoldmopar commented:
Alright, this is looking great.  I pulled in develop one more time and did a full build and test.  Unit and integration tests are passing fine, and there are no unexpected regressions.  This is good to go in.  I'll push my small commit up and then get it merged.  Thanks @j-c-cook and @mitchute !

Myoldmopar committed:
Address Windows build warnings

Myoldmopar commented:
Forgot about the Windows warnings, so addressed them real quick.  Heads up @j-c-cook these changes were in the new third party code.

Myoldmopar commented on third_party/cpgfunctionEP-0.5/src/segments.cpp:
```diff
@@ -49,7 +49,7 @@ namespace gt::segments {
             nq = jcc::interpolation::interp1d(drilling_depth,
                                               drilling_depths[0],
                                               ideal_segment_lengths[0]);
-        } else if(heights[0] < height < heights[n1]) {
```
In C++, `x < y < z` doesn't do what you think it does.  C++ does not have chained operation.  This will perform one binary comparison, which in this case returns a boolean, then try to compare that boolean to a floating point with the second binary operation.  I'm assuming you just want to do `x < y && y < z`, so that's what I did here.  This nifty syntax works in Python but not C++.

Myoldmopar commented on third_party/cpgfunctionEP-0.5/third_party/LU-Decomposition-0.1/src/lu.cpp:
```diff
@@ -9,7 +9,7 @@ void jcc::decomposition(vector<vector<double> > &A, int &n, vector<int> &indx,
     const double TINY = 1.0e-20;  // a small number
     int i, imax, j, k;
-    double big, dum, sum, temp;
+    double big, dum, sum;
```
Unused variable removd.

Myoldmopar commented on third_party/cpgfunctionEP-0.5/third_party/qdt-master/methods/adaptive.h:
```diff
@@ -376,7 +376,7 @@ class Adaptive<BaseMethod,Estimator,true,true,false,true> :
 Adaptive<BM> adaptive(const BM& bm)
 {
-	return Adaptive<BM>(bm, 1.e-5, 1.e-10);
+	return Adaptive<BM>(bm, 1.e-5f, 1.e-10f);
```
Explicitly creating float literals instead of double to avoid the warning.

Myoldmopar commented:
Alright, I'm merging this now.  Thanks all.

rraustad commented:
@j-c-cook why do I have to add #include \<algorithm\> to boreholes.cpp and adaptive.h to get MS Visual Studio Professional 2019 Version 16.5.0 to compile? In boreholes.cpp was a max statement where MSVS said identifier max not found and in adaptive.h is a std::max and MSVS says max is not a member. When I type in std::m I don't see max in the pop up list as an available function. I also don't understand why adding #include \<algorithm\> in adaptive.h corrects a statement of std::max ?? since I did not change that line from std::max to max.

Do you have any insight into this?

![image](https://user-images.githubusercontent.com/5778156/130288977-4de03eaf-99a7-4dd3-b522-4be9de560df5.png)

rraustad commented:
In develop boreholes.cpp I see this in MSVS.

![image](https://user-images.githubusercontent.com/5778156/130292164-87a11bfb-b1af-458a-95d9-0c2915680795.png)

j-c-cook commented:
I cannot say anything definitive about how MS Visual Studio Professional 2019 Version 16.5.0 operates. I do nearly all development in Kubuntu 20.04.2, and compiled the code in Windows 10 only when the CI machine returned errors. I was under the impression that all errors in a Windows build or test would have been caught by the CI machines, and that the code would not have been merged before those errors were found. However, unforeseen bugs happen. 

To eradicate this bug will I need to download a Visual Studio IDE specifically, or does compiling from the command prompt provide the same error?

rraustad commented:
I can just commit this small change as it does not appear to change anything other than which headers are included. I just wanted to ask before I made that commit. It's only 2 lines of code, 1 in each file, and only a #include. Why I see this compiler warning and others who use MSVS do not is a mystery to me (other than they haven't pulled develop yet).

j-c-cook commented:
Okay thank you! I'll make sure those changes propagate their way back into cpgfunctionEP's repo, along with some refactors Edwin had made as well.
